File: src/adapters.ts
import { spawnSync } from "child_process";
import { join } from "path";
import { fail } from "./errors";
import { appendEvent, atomicWriteFile, ensureDir, goalIntegrity, nowIso, readAdapters, readPolicy } from "./io";
import {
  CRITIC_PROMPT_PATH,
  DECISIONS_PATH,
  FAILURES_PATH,
  GOAL_PATH,
  PLAN_PATH,
  POLICY_PATH,
  PROGRESS_PATH,
  RAW_DIR,
  RUNS_DIR,
  TASKS_PATH,
  WORKER_PROMPT_PATH,
} from "./paths";
import { AgentAdapter, AgentRole } from "./schemas";

export function runAdapter(root: string, adapterId: string, role: AgentRole, execute: boolean): void {
  goalIntegrity(root);
  const policy = readPolicy(root);
  const { adapter, args, renderedCommand } = resolveAdapterCommand(root, adapterId, role);
  if (!execute) {
    console.log(`dry run: ${renderedCommand}`);
    return;
  }

  ensureDir(join(root, RUNS_DIR));
  const runId = `${adapter.id}-${role}-${compactTimestamp(nowIso())}-${Math.random().toString(16).slice(2, 10)}`;
  appendEvent(root, {
    timestamp: nowIso(),
    kind: "adapter.run_started",
    actor: "yoloop",
    taskId: null,
    message: `Started adapter ${adapter.id} as ${role}.`,
    data: { adapter: adapter.id, role, command: renderedCommand, runId },
  });

  const output = spawnSync(adapter.command, args, {
    cwd: root,
    encoding: "utf8",
    shell: false,
    timeout: policy.maxWallClockMinutes * 60 * 1000,
  });

  const stdoutPath = join(root, RUNS_DIR, `${runId}.stdout.txt`);
  const stderrPath = join(root, RUNS_DIR, `${runId}.stderr.txt`);
  atomicWriteFile(stdoutPath, output.stdout ?? "");
  atomicWriteFile(stderrPath, output.stderr ?? "");

  appendEvent(root, {
    timestamp: nowIso(),
    kind: "adapter.run_finished",
    actor: "yoloop",
    taskId: null,
    message: `Finished adapter ${adapter.id} as ${role}.`,
    data: {
      adapter: adapter.id,
      role,
      runId,
      exitCode: output.status,
      signal: output.signal,
      error: output.error?.message,
    },
  });

  if (output.error) {
    fail(`adapter failed: ${output.error.message}`);
  }
  if (output.status !== 0) {
    fail(`adapter exited with status ${output.status}`);
  }
}

export function resolveAdapterCommand(
  root: string,
  adapterId: string,
  role: AgentRole,
): { adapter: AgentAdapter; args: string[]; renderedCommand: string } {
  const catalog = readAdapters(root);
  const adapter = catalog.adapters.find((candidate) => candidate.id === adapterId);
  if (!adapter) {
    fail(`adapter ${adapterId} not found in ADAPTERS.json`);
  }
  const args = renderAdapterArgs(adapter, role);
  return { adapter, args, renderedCommand: formatCommand(adapter.command, args) };
}

export function renderAdapterArgs(adapter: AgentAdapter, role: AgentRole): string[] {
  const args =
    role === "worker"
      ? adapter.workerArgs
      : role === "critic"
        ? adapter.criticArgs
        : adapter.grandJuryArgs;
  return args.map(renderTemplate);
}

export function renderTemplate(value: string): string {
  return value
    .replace(/\{\{goal\}\}/g, GOAL_PATH)
    .replace(/\{\{policy\}\}/g, POLICY_PATH)
    .replace(/\{\{plan\}\}/g, PLAN_PATH)
    .replace(/\{\{tasks\}\}/g, TASKS_PATH)
    .replace(/\{\{worker_prompt\}\}/g, WORKER_PROMPT_PATH)
    .replace(/\{\{critic_prompt\}\}/g, CRITIC_PROMPT_PATH)
    .replace(/\{\{progress\}\}/g, PROGRESS_PATH)
    .replace(/\{\{failures\}\}/g, FAILURES_PATH)
    .replace(/\{\{decisions\}\}/g, DECISIONS_PATH)
    .replace(/\{\{raw\}\}/g, RAW_DIR);
}

function formatCommand(command: string, args: string[]): string {
  return [command, ...args].map(shellQuote).join(" ");
}

function shellQuote(value: string): string {
  if (/^[A-Za-z0-9_.\\/:/-]+$/.test(value)) {
    return value;
  }
  return `"${value.replace(/"/g, '\\"')}"`;
}

function compactTimestamp(iso: string): string {
  return iso.replace(/[-:]/g, "").replace(".", "");
}

File: src/adapters.rs
use crate::{
    fsio::{append_event, ensure_dir, goal_integrity, read_adapters},
    models::{AgentAdapter, AgentRole, Event},
    paths::{
        ADAPTERS_PATH, CRITIC_PROMPT_PATH, DECISIONS_PATH, FAILURES_PATH, GOAL_PATH, PLAN_PATH,
        POLICY_PATH, PROGRESS_PATH, RAW_DIR, RUNS_DIR, TASKS_PATH, WORKER_PROMPT_PATH,
    },
};
use anyhow::{bail, Context, Result};
use chrono::Utc;
use serde_json::json;
use std::{collections::BTreeMap, fs, path::Path, process::Command as ProcessCommand};

pub(crate) fn run_adapter(
    root: &Path,
    adapter_id: &str,
    role: AgentRole,
    execute: bool,
) -> Result<()> {
    goal_integrity(root)?;
    let catalog = read_adapters(root)?;
    let adapter = catalog
        .adapters
        .iter()
        .find(|candidate| candidate.id == adapter_id)
        .with_context(|| format!("adapter {adapter_id} not found in {ADAPTERS_PATH}"))?;
    let args = render_adapter_args(adapter, role);
    let rendered_command = format_command(&adapter.command, &args);

    if !execute {
        println!("dry run: {rendered_command}");
        return Ok(());
    }

    ensure_dir(root.join(RUNS_DIR))?;
    let run_id = format!(
        "{}-{}-{}",
        adapter.id,
        role.as_str(),
        Utc::now().format("%Y%m%dT%H%M%SZ")
    );

    let mut data = BTreeMap::new();
    data.insert("adapter".to_string(), json!(adapter.id));
    data.insert("role".to_string(), json!(role.as_str()));
    data.insert("command".to_string(), json!(rendered_command));
    data.insert("runId".to_string(), json!(run_id));
    append_event(
        root,
        Event {
            timestamp: Utc::now(),
            kind: "adapter.run_started".to_string(),
            actor: "yoloop".to_string(),
            task_id: None,
            message: format!("Started adapter {} as {}.", adapter.id, role.as_str()),
            data,
        },
    )?;

    let output = ProcessCommand::new(&adapter.command)
        .args(&args)
        .current_dir(root)
        .output()
        .with_context(|| format!("run {rendered_command}"))?;

    let stdout_path = root.join(RUNS_DIR).join(format!("{run_id}.stdout.txt"));
    let stderr_path = root.join(RUNS_DIR).join(format!("{run_id}.stderr.txt"));
    fs::write(&stdout_path, output.stdout)
        .with_context(|| format!("write {}", stdout_path.display()))?;
    fs::write(&stderr_path, output.stderr)
        .with_context(|| format!("write {}", stderr_path.display()))?;

    let mut data = BTreeMap::new();
    data.insert("adapter".to_string(), json!(adapter.id));
    data.insert("role".to_string(), json!(role.as_str()));
    data.insert("runId".to_string(), json!(run_id));
    data.insert("exitCode".to_string(), json!(output.status.code()));
    append_event(
        root,
        Event {
            timestamp: Utc::now(),
            kind: "adapter.run_finished".to_string(),
            actor: "yoloop".to_string(),
            task_id: None,
            message: format!("Finished adapter {} as {}.", adapter.id, role.as_str()),
            data,
        },
    )?;

    if !output.status.success() {
        bail!("adapter exited with status {}", output.status);
    }

    Ok(())
}

fn render_adapter_args(adapter: &AgentAdapter, role: AgentRole) -> Vec<String> {
    let args = match role {
        AgentRole::Worker => &adapter.worker_args,
        AgentRole::Critic => &adapter.critic_args,
        AgentRole::GrandJury => &adapter.grand_jury_args,
    };
    args.iter().map(|arg| render_template(arg)).collect()
}

fn render_template(value: &str) -> String {
    value
        .replace("{{goal}}", GOAL_PATH)
        .replace("{{policy}}", POLICY_PATH)
        .replace("{{plan}}", PLAN_PATH)
        .replace("{{tasks}}", TASKS_PATH)
        .replace("{{worker_prompt}}", WORKER_PROMPT_PATH)
        .replace("{{critic_prompt}}", CRITIC_PROMPT_PATH)
        .replace("{{progress}}", PROGRESS_PATH)
        .replace("{{failures}}", FAILURES_PATH)
        .replace("{{decisions}}", DECISIONS_PATH)
        .replace("{{raw}}", RAW_DIR)
}

fn format_command(command: &str, args: &[String]) -> String {
    std::iter::once(shell_quote(command))
        .chain(args.iter().map(|arg| shell_quote(arg)))
        .collect::<Vec<_>>()
        .join(" ")
}

fn shell_quote(value: &str) -> String {
    if value
        .chars()
        .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '-' | '_' | '.' | '/' | '\\' | ':'))
    {
        value.to_string()
    } else {
        format!("\"{}\"", value.replace('"', "\\\""))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_raw_placeholder() {
        assert_eq!(
            render_template("read {{raw}} and {{goal}}"),
            "read raw and GOAL.html"
        );
    }
}
